engines/immortal/kernal.cpp

```diff
@@ -0,0 +1,1104 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/* This file covers both Kernal.GS and Driver.GS.
+ * This is because most of Driver.GS is hardware specific,
+ * and what is not (the slightly abstracted aspects), is
+ * directly connected to kernal, and might as well be
+ * considered part of the same process.
+ */
+
+#include "immortal/immortal.h"
+
+namespace Immortal {
+
+/*
+ *
+ * -----                          -----
+ * ----- Screen Drawing Functions -----
+ * -----                          -----
+ *
+ */
+
+void ImmortalEngine::drawUniv() {
+	// This is where the entire screen actually comes together
+	_myViewPortX = _viewPortX & 0xFFFE;
+	_myViewPortY = _viewPortY;
+
+	_num2DrawItems = 0;
+
+	_myUnivPointX = !(_myViewPortX & (kChrW - 1)) + kViewPortSpX;
+	_myUnivPointY = !(_myViewPortY & (kChrH - 1)) + kViewPortSpY;
+
+	//makeMyCNM();
+	//drawBGRND();                          // Draw floor parts of leftmask rightmask and maskers
+	//addRows();                                // Add rows to drawitem array
+	//addSprites();                         // Add all active sprites that are in the viewport, into a list that will be sorted by priority
+	//sortDrawItems();                      // Sort said items
+	//drawItems();                          // Draw the items over the background
+}
+
+void ImmortalEngine::copyToScreen() {
+	/* copyRectToSurface will apply the screenbuffer to the ScummVM surface.
+	 * We want to do 320 bytes per scanline, at location (0,0), with a
+	 * size of 320x200.
+	 */
+	_mainSurface->copyRectToSurface(_screenBuff, kResH, 0, 0, kResH, kResV);
```

Comment:
If you're copying the whole screen every time, I'd suggest using Dirty rectangles.